### ISSUE TITLE
Parsing TDH_INTYPE_WBEMSID

### ISSUE THREAD

**Issue by FuzzySecurity** [microsoft/krabsetw#70]
I created a tiny wrapper for a number of ETW events, however there are some `TDH_IN_TYPE` types which I am unable to parse. For example:

![image](https://user-images.githubusercontent.com/10997340/52812360-2de5c400-308f-11e9-872d-fbd70aa915ec.png)

Here type 310 is `TDH_INTYPE_WBEMSID` which should be a `TOKEN_USER` struct followed by a `SID` but I can't find a way to access the value of this property. I have a `switch(prop.Type)->case` statement to handle the different types, eg:

```c#
case (int)TDH_IN_TYPE.TDH_INTYPE_UINT32:
    propertyValue = record.GetUInt32(prop.Name);
    break;
```

If I can cast the result as an `IntPtr` I can make it work: Increment the pointer by the size of `TOKEN_USER` and then call `LookupAccountSid` etc. However I can't make it work, if I try something like this I get an exception saying the requested size does not match the size of the parameter.

```c#
case (int)TDH_IN_TYPE.TDH_INTYPE_WBEMSID:
    IntPtr pRecod = (IntPtr)record.GetUInt64(prop.Name );
    break;
```

Thanks for the great wrapper!

**Comment by zacbrown** [comment]
Hi @FuzzySecurity,

Thanks for the feedback. Yes - there are a number of types not completely supported in krabsetw. If you're willing to create a PR with the fix, we're happy to accept it.

Otherwise, the issue will get fixed as either the engineering team at MSFT gets to it if I happen to find some free time :).

Thanks!

-Zac

**Comment by FuzzySecurity** [comment]
I can implement the `WBEMSID` parsing but my question is how can I access the raw value of the property? The return type here should be IntPtr right?

**Comment by zacbrown** [comment]
Ah - you'd need to update the libraries to support a pointer type. While we do support interpreting the `TDH_INTYPE_WBEMSID` size type, there's no declaration of its size type in here: https://github.com/Microsoft/krabsetw/blob/0b42d12c337d2ea954ebc89c1547be48a552cf13/krabs/krabs/tdh_helpers.hpp#L153

That would require someone spending the time identifying the proper way to access the members -as you're currently experiencing.

Right now, krabsetw and its .NET wrapper don't support pointer types besides the assorted string types. When the library was originally created, it was not something we had an immediate need for IIRC though @swannman may be able to confirm that.

**Comment by swannman** [comment]
I would definitely accept pull requests that add support for new types (whether pointer or value types). I think the PR would need to allocate the appropriate struct on the stack, follow the pointer to do a deep-copy of the value members into the struct, and return the struct to the caller. Is that accurate?

@FuzzySecurity for an example of adding support for `WBEMSID`, grep for `TDH_OUTTYPE_SOCKETADDRESS` and `socket_address`. Here's roughly what would be involved: 

1. define an appropriate struct in `parse_types.hpp`
2. add a mapping from the TDH type to your struct in `filler.hpp`
3. define a `parser::parse()` overload in `parser.hpp`
4. add an appropriate debug assert at `tdh_helpers.hpp`
5. and implement an appropriate C++/CLI wrapper in `EventRecord.hpp`

This looks pretty achievable - if you'd be willing to submit a PR, we'll take a look and help you get it into shape.

Cheers,
Matt

**Comment by manishshukla** [comment]
A possible alternative to is to use `krabs::binary` for getting the raw data. It is much easier than modifying krabsetw internal files. Code snippet used:

```
krabs::schema sc(record);
krabs::parser pa(sc);
krabs::binary bin;

/*
    As suggested here:
    https://docs.microsoft.com/en-us/windows/desktop/etw/using-tdhgetproperty-to-consume-event-data
*/
auto pointer_size = 8;
if (EVENT_HEADER_FLAG_32_BIT_HEADER == (record.EventHeader.Flags & EVENT_HEADER_FLAG_32_BIT_HEADER))
	pointer_size = 4;

auto start_point = pointer_size * 2;

switch(opcode) {
    case EVENT_TRACE_TYPE_START: 
    bin = pa.parse<krabs::binary>(L"UserSID"); 
    {
        WCHAR user_name[MAX_NAME]; // #define MAX_NAME 256
		WCHAR domain_name[MAX_NAME];
		DWORD user_name_size = MAX_NAME;
		DWORD domain_name_size = MAX_NAME;
		SID_NAME_USE name_use;
		const BYTE* data = (sid.bytes().data() + start_point);
		if (LookupAccountSid(NULL, (PSID)data, user_name, &user_name_size, domain_name, &domain_name_size, &name_use)) {
			std::wcout << domain_name << L"\\" << user_name << L"\n";
		}
    }
    break;
}
```

**Comment by manishshukla** [comment]
An alternative is to use `krabs::binary` for getting the raw data and then calling the appropriate API for further processing. It is much easier than modifying the krabsetw internal code. Minimal code snippet:

```
krabs::schema sc(record);

/*
    As suggested here:
    https://docs.microsoft.com/en-us/windows/desktop/etw/using-tdhgetproperty-to-consume-event-data
*/
auto pointer_size = 8;
if (EVENT_HEADER_FLAG_32_BIT_HEADER == (sc.event_flags() & EVENT_HEADER_FLAG_32_BIT_HEADER))
	pointer_size = 4;

auto start_point = pointer_size * 2;
				
krabs::parser pa(sc);
krabs::binary sid;

switch(sc.event_opcode()) {
    case EVENT_TRACE_TYPE_START: 
    sid = pa.parse<krabs::binary>(L"UserSID");
    {
        wchar_t user_name[MAX_NAME] = L""; // #define MAX_NAME 256
	wchar_t domain_name[MAX_NAME] = L"";
	DWORD user_name_size = MAX_NAME;
	DWORD domain_name_size = MAX_NAME;
	SID_NAME_USE name_use;
	const BYTE* data = (sid.bytes().data() + start_point);
	if (LookupAccountSid(NULL, (PSID)data, user_name, &user_name_size, domain_name, &domain_name_size, &name_use)) {
		std::wcout << domain_name << L"\\" << user_name << L"\n";
	}
    }
    break;
}
```

**Comment by pathtofile** [comment]
So I also came across this 'issue', so I had a shot at implementing it. As I mention in the PR, I made two major assumptions that if you don't agree with I can fix:

1) In `TDH_INTYPE_WBEMSID` we only care about the SID part, not the `TOKEN_USER` - This is based upon this MSDN example, which also only pulls out the SID: https://docs.microsoft.com/en-us/windows/win32/etw/using-tdhgetproperty-to-consume-event-data. Also as this type is depracted, and you're meant to only use TDH_INTYPE_SID. This meant we can 'decode' both TDHs to the same struct.

2) I store the string of the SID instead of the bytes. As SIDs are variable-length, this seemed like the easiest way to deal with them, as we can have the string on the stack instead of allocating and copying the varaible-length bytes.
